giraffe/src/components/AnnotationLayer.tsx

```diff
@@ -0,0 +1,47 @@
+import React, {FunctionComponent, useMemo} from 'react'
+import {AnnotationLayerConfig, AnnotationLayerSpec, LayerProps} from '../types'
+import {getAnnotationsPositions} from '../utils/annotationData'
+
+interface AnnotationLayerProps extends LayerProps {
+  spec: AnnotationLayerSpec
+  config: AnnotationLayerConfig
+}
+
+const ANNOTATION_OVERLAY_DEFAULT_STYLE = {
+  width: '100%',
+  height: '100%',
+}
+
+export const AnnotationLayer: FunctionComponent<AnnotationLayerProps> = props => {
```

Comment:
philosophical question: why a `FunctionComponent` over a regular component or a `PureComponent`? i'm not sure if this has changed in the past couple of versions, but my understanding is that a `PureComponent` will only re-render when its props or state changes, whereas a `FunctionComponent` will always re-render and we have to use a fairly complex concept like memoization to make it performant.

Reply:
- The way that AnnotationLayer is used, the props will have changed when it re-renders. The props that change are the mouse coordinates, and they will change if the mouse is hovering over any part of the graph (or AnnotationLayer).

- So, from a re-rendering perspective, it doesn't matter if AnnotationLayer is a function component or a class component. But there is no advantage in being a class component in our use case. Whereas, there are benefits for being a function component. See https://medium.com/@Zwenza/functional-vs-class-components-in-react-231e3fbd7108 under "So why should I use functional components at all?" Additionally, item 4 has a link that leads to a React release blog that mentions the performance impact.

- The memoization is a benefit regardless of whether this is a function component or class component. Even written as a class component, memoization should be done for the positions of the annotation marks.

- And ultimately, i think "re-render" means that the function component is executed, ie the function is called. Actual changes to the DOM is where the performance bottleneck happens. Here, we rely on React to diff and update.

Reply:
Without getting into a discussion about individual perferences, that blog post's justifications to use functional components were all based on opinion. In some cases, functional components are easier to test. In other cases, they're much harder to test. In some cases they're less code, in other cases they're more code. And as far as I know, the React team hasn't done anything towards functional component performance updates in the two years since they made those announcements. I think they've backed off the "functions only" train a bit. The point is, it's preference - there are no sound technical reasons to always use one or the other. Sometimes functional components make more sense, sometimes class components and the structure they have make more sense. I bring it up not to say don't use an FC here, but to say 'beware arguments like "the code is cleaner this way"' and also to say that I do not want us to get locked into the mindset that functional components are the only way forward for all our new code.

Reply:
Very valid points.

Yes, i do agree that we should look at this from a perspective that is outside of any personal preferences.

My understanding is that class components are almost always more code, because an instance has to be created from a parent class component. In cases where a function component has to rely on a lot of hooks, then maybe the two are closer in terms of "size". But that is probably a case where the component can be broken down further. When a component is reasonably small, I see the functional component as having less code.

But I am certainly aware that there's often a better way to do something. If there is a case where a class component is better at the job, then we'll go with the class component.

I just think that a mindset of defaulting to using class components is not good either. Since that is often a developer's first exposure to React, they tend to fall back on what is comfortable--class components.